pkg/network/ebpf/c/protocol-classification-helpers.h

```diff
@@ -0,0 +1,207 @@
+#ifndef __PROTOCOL_CLASSIFICATION_HELPERS_H
+#define __PROTOCOL_CLASSIFICATION_HELPERS_H
+
+#include <linux/types.h>
+
+#include "protocol-classification-defs.h"
+#include "protocol-classification-maps.h"
+#include "ip.h"
+
+// Patch to support old kernels that does not support bpf_skb_load_bytes, by adding a dummy implementation.
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 5, 0)
+long bpf_skb_load_bytes(const void *skb, u32 offset, void *to, u32 len) {return 0;}
+#endif
+
+// The method checks if the given buffer starts with the HTTP2 marker as defined in https://datatracker.ietf.org/doc/html/rfc7540.
+// We check that the given buffer is not empty and its size is at least 24 bytes.
+static __always_inline bool is_http2(const char* buf, __u32 buf_size) {
+    if (buf_size < HTTP2_MARKER_SIZE) {
+        return false;
+    }
+
+    if (buf == NULL) {
+        return false;
+    }
+
+    // Unfortunately, the compiler tries to outsmart and causes the verifier on older kernels to think we have more
+    // than a millions possible instructions in the code and thus it fails to verify and load it.
+    // Using volatile parameter to tell the compiler to not optimize the code.
+    volatile bool match = buf[0]=='P' && buf[1]=='R' && buf[2]=='I' && buf[3]==' ' && buf[4]=='*' && buf[5]==' ' &&
+        buf[6]=='H' && buf[7]=='T' && buf[8]=='T' && buf[9]=='P' && buf[10]=='/' && buf[11]=='2' && buf[12]=='.' &&
+        buf[13]=='0' && buf[14]=='\r' && buf[15]=='\n' && buf[16]=='\r' && buf[17]=='\n' && buf[18]=='S' &&
+        buf[19]=='M' && buf[20]=='\r' && buf[21]=='\n' && buf[22]=='\r' && buf[23]=='\n';
+    return match;
+}
+
+// Checks if the given buffers start with `HTTP` prefix (represents a response) or starts with `<method> /` which represents
+// a request, where <method> is one of: GET, POST, PUT, DELETE, HEAD, OPTIONS, or PATCH.
+static __always_inline bool is_http(const char *buf, __u32 size) {
+    if (size < HTTP_MIN_SIZE) {
+        return false;
+    }
+    if (buf == NULL) {
+        return false;
+    }
+
+    if ((buf[0] == 'H') && (buf[1] == 'T') && (buf[2] == 'T') && (buf[3] == 'P')) {
```

Comment:
so that's the original code we are using in USM ( i've just moved it around)

but eventually, the scenario you are mentioning will always occur if someone sends `GET /not HTTP` over a TCP message we will have no way but to say that's HTTP because you cannot know

for every hardening of the check I can find how to overcome the check